File: trunk/sources/net.sf.j2s.java.org.eclipse.osgi/src/org/eclipse/core/runtime/internal/adaptor/Semaphore.java
/*******************************************************************************
 * Copyright (c) 2003, 2004 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package org.eclipse.core.runtime.internal.adaptor;

/**
 * Internal class.
 */
public class Semaphore {
	protected long notifications;

	public Semaphore(int count) {
		notifications = count;
	}

	/**
	 * Attempts to acquire this semaphore.  Returns only when the semaphore has been acquired.
	 */
	public synchronized void acquire() {
		while (true) {
			if (notifications > 0) {
				notifications--;
				return;
			}
			try {
				wait();
			} catch (InterruptedException e) {
				//Ignore
			}
		}
	}

	/**
	 * Attempts to acquire this semaphore.  Returns true if it was successfully acquired,
	 * and false otherwise.
	 */
	public synchronized boolean acquire(long delay) {
		long start = System.currentTimeMillis();
		long timeLeft = delay;
		while (true) {
			if (notifications > 0) {
				notifications--;
				return true;
			}
			if (timeLeft < 0)
				return false;
			try {
				wait(timeLeft);
			} catch (InterruptedException e) {
				//Ignore
			}
			timeLeft = start + delay - System.currentTimeMillis();
		}
	}

	public synchronized void release() {
		notifications++;
		notifyAll();
	}

	// for debug only
	public String toString() {
		return "Semaphore(" + notifications + ")"; //$NON-NLS-1$ //$NON-NLS-2$
	}
}

File: old/incubator/net.sf.j2s.test.mauve/j2slib/org/eclipse/core/runtime/internal/adaptor/Semaphore.js
Clazz.declarePackage ("org.eclipse.core.runtime.internal.adaptor");
c$ = Clazz.decorateAsClass (function () {
this.notifications = 0;
Clazz.instantialize (this, arguments);
}, org.eclipse.core.runtime.internal.adaptor, "Semaphore");
Clazz.makeConstructor (c$, 
function (count) {
this.notifications = count;
}, "~N");
Clazz.defineMethod (c$, "acquire", 
function () {
while (true) {
if (this.notifications > 0) {
this.notifications--;
return ;
}try {
this.wait ();
} catch (e) {
if (Clazz.instanceOf (e, InterruptedException)) {
} else {
throw e;
}
}
}
});
Clazz.defineMethod (c$, "acquire", 
function (delay) {
var start = System.currentTimeMillis ();
var timeLeft = delay;
while (true) {
if (this.notifications > 0) {
this.notifications--;
return true;
}if (timeLeft < 0) return false;
try {
this.wait (timeLeft);
} catch (e) {
if (Clazz.instanceOf (e, InterruptedException)) {
} else {
throw e;
}
}
timeLeft = start + delay - System.currentTimeMillis ();
}
}, "~N");
Clazz.defineMethod (c$, "release", 
function () {
this.notifications++;
this.notifyAll ();
});
Clazz.overrideMethod (c$, "toString", 
function () {
return "Semaphore(" + this.notifications + ")";
});
